motion/colav/scripts/VOMATH.py

```diff
@@ -0,0 +1,302 @@
+        self.x = x        
+        self.y = y        
+        self.vx = vx        
+        self.vy = vy        
+        self.radius = radius    
+    
+    def update(self, x, y, vx, vy):
+        self.x = x        
+        self.y = y        
+        self.vx = vx        
+        self.vy = vy
+
+def vo_collision_avoidance(ownship, intruders):
+    # Compute velocity obstacles for each intruder    
+    v_obs = []
+    for intruder in intruders:
+        r = ownship.radius + intruder.radius + VO_MARGIN        
+        u = intruder.vx - ownship.vx        
+        v = intruder.vy - ownship.vy        
+        u1 = u + r * np.sign(u) * np.sqrt(u**2 + v**2 + MAX_SPEED**2)
+        u2 = u - r * np.sign(u) * np.sqrt(u**2 + v**2 + MAX_SPEED**2)
+        v1 = v + r * np.sign(v) * np.sqrt(u**2 + v**2 + MAX_SPEED**2)
+        v2 = v - r * np.sign(v) * np.sqrt(u**2 + v**2 + MAX_SPEED**2)
+        v_obs.append((u1, u2, v1, v2))
+
+    # Compute the reachable set of velocities for the ownship    
+    vx_reach = np.arange(-MAX_SPEED, MAX_SPEED + 0.1, 0.1)
+    vy_reach = np.arange(-MAX_SPEED, MAX_SPEED + 0.1, 0.1)
+    Vx, Vy = np.meshgrid(vx_reach, vy_reach)
+    Vx = Vx.flatten()
+    Vy = Vy.flatten()
+    V = np.array([Vx, Vy]).T    
+    V_reach = []
+    
+    for v in V:
+        if np.linalg.norm(v) < MAX_SPEED:
+            a = np.array([MAX_ACCELERATION, 0])
+            b = np.array([v[0], v[1]])
+            if np.arccos(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))) < MAX_TURN_RATE:
+                V_reach.append(v)
+    V_reach = np.array(V_reach)
+    # Find the velocity that maximizes the distance to the velocity obstacles    
+    # Må endres
+    best_v = np.array([0, 0])
+    best_dist = -np.inf    
+    for v in V_reach:
+        dists = []
+        for obs in v_obs:
+            if v[0] > obs[0] or v[0] < obs[1] or v[1] > obs[2] or v[1] < obs[3]:
+                dists.append(0)
+            else:
+                d = np.min([obs[0] - v[0], v[0] - obs[1], obs[2] - v[1], v[1] - obs[3]])
+                dists.append(d)
+        dist = np.min(dists)
+        if dist > best_dist:
+                    # Check if the chosen velocity is valid          
+            a = best_v - np.array([ownship.vx, ownship.vy])
+            b = v - np.array([ownship.vx, ownship.vy])
+            if np.arccos(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))) > MAX_TURN_RATE:
+                continue          
+            ownship.update(ownship.x + v[0], ownship.y + v[1], v[0], v[1])
+            return ownship      
+    return None
+
+def spin(self):
+    self.enable = rospy.get_param("/")
+
+    while True:
+        if self.enable:
+            # for all obstacles:
+            # check if could have collison, red, green yellow
+            # if green do nothing
+            # if yellow give new heading
+            # if red, stop!
+            pass 
+        # TODO: 
+        #implemen tmath
+        #implement landmar server
```

Comment:
Complete the function or explain why it does not do anything (a TODO?), or remove if it is not needed